### PR TITLE
plot partial dependence for symmetric trees

### PR DESCRIPTION
**Pull request by felixandrer**
New plot_partial_dependence function to plot PDP (https://christophm.github.io/interpretable-ml-book/pdp.html) for symmetric trees.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en.

### CODE REVIEW

File: catboost/libs/fstr/partial_dependence.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
+#include "partial_dependence.h"
+
+#include <catboost/private/libs/algo/apply.h>
+#include <catboost/libs/helpers/exception.h>
+#include <catboost/libs/model/model.h>
+#include <catboost/libs/data/data_provider.h>
+#include <util/generic/utility.h>
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:7]: blank line between catboost include block and util include block

File: catboost/libs/fstr/partial_dependence.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
+#include "partial_dependence.h"
+
+#include <catboost/private/libs/algo/apply.h>
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:3]: sort lines in catboost/... include block in lexicographical order

File: catboost/libs/fstr/partial_dependence.cpp [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+#include <catboost/libs/model/model.h>
+#include <catboost/libs/data/data_provider.h>
+#include <util/generic/utility.h>
+
+
+using namespace NCB;
+
+namespace {
+    struct TFloatFeatureBucketRange {
+        int featureIdx = -1;
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:14]: fields in classes are written in CamelCase

File: catboost/libs/fstr/partial_dependence.cpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+        TFloatFeatureBucketRange(int featureIdx, int numOfBorders)
+                : featureIdx(featureIdx)
+                , start(0)
+                , end(numOfBorders + 1)
+                , numOfBuckets(numOfBorders + 1)
+        {
+        }
+
+        void update(int borderIdx, bool isGreater) {
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:29]: function names must be written in CamelCase

File: catboost/libs/fstr/partial_dependence.cpp [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+TVector<TFloatFeatureBucketRange> prepareFeatureRanges(
+        const TFullModel& model,
+        const TVector<int>& featuresIdx
+) {
+    TVector<TFloatFeatureBucketRange> featureRanges(0);
+    if (featuresIdx.size() < 2) {
+        featureRanges.push_back(TFloatFeatureBucketRange(-1, 0));
+    }
+    for (int idx: featuresIdx) {
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:47]: space before colon

File: catboost/libs/fstr/partial_dependence.cpp [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        const auto& feature = model.ModelTrees->GetFloatFeatures()[idx];
+        const auto& range = TFloatFeatureBucketRange(feature.Position.Index, feature.Borders.size());
+        featureRanges.push_back(range);
+    }
+    return featureRanges;
+}
+
+void UpdateFeatureRanges(
+        const TFullModel& model,
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:56]: 4 spaces indent

File: catboost/libs/fstr/partial_dependence.cpp [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        TVector<TFloatFeatureBucketRange>* featureRanges,
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:57]: non const argument should be passed after const args

File: catboost/libs/fstr/partial_dependence.cpp [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            if (range.featureIdx == split.FloatFeature.FloatFeature) {
+                int borderIdx = borderIdxForSplit[splits[splitIdx]];
+                range.update(borderIdx, decision);
+            }
+        }
+    }
+
+}
+
+std::pair<TVector<TVector<TFloatFeatureBucketRange>>, TVector<double>>  CalculateBucketRangesAndWeightsOblivious(
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:77]: 2 spaces

File: catboost/libs/fstr/partial_dependence.cpp [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+TVector<double> MergeBucketRanges(
+        const TFullModel& model,
+        const TVector<int>& features,
+        const TDataProvider& dataProvider,
+        const TVector<TVector<TFloatFeatureBucketRange>>& leafBucketRanges,
+        const TVector<double> leafWeights
+) {
+    const auto& leafValues = model.ModelTrees->GetLeafValues();
+    TVector<TFloatFeatureBucketRange> defaultRanges = prepareFeatureRanges(model, features);
+    CB_ENSURE(defaultRanges.size() == 2, "Number of features must be 2");
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:141]: this error message is not clear

File: catboost/libs/fstr/partial_dependence.cpp [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+                    edges[rowIdx * columnNum + ranges[1].end] -= leafValue * leafWeights[leafIdx];
+                } else if (rowIdx < rowNum - 1) {
+                    edges[(rowIdx + 1) * columnNum] -= leafValue * leafWeights[leafIdx];
+                }
+            }
+        }
+    }
+
+    TVector<double> predictionsByBuckets(numOfBucketsTotal);
+    double acc = 0;
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:164]: an unintelligible variable name

File: catboost/libs/fstr/partial_dependence.cpp [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+
+    const auto& binSplits = model.ModelTrees->GetBinFeatures();
+
+    TVector<ui32> borderIdxForSplit(binSplits.size(), std::numeric_limits<ui32>::infinity());
+    ui32 splitIdx = 0;
+    for (const auto& feature : model.ModelTrees->GetFloatFeatures()) {
+        if (splitIdx == binSplits.size() ||
+            binSplits[splitIdx].Type != ESplitType::FloatFeature ||
+            binSplits[splitIdx].FloatFeature.FloatFeature > feature.Position.Index
+                ) {
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:221]: here are 2 extra indents

File: catboost/libs/fstr/partial_dependence.cpp [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+            continue;
+        }
+        Y_ASSERT(binSplits[splitIdx].FloatFeature.FloatFeature >= feature.Position.Index);
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.cpp:224]: Y_ASSERT is for debug, use CB_ENSURE_INTERNAL instead

File: catboost/libs/fstr/partial_dependence.h [hunk 13]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <catboost/libs/data/data_provider.h>
+#include <catboost/libs/model/model.h>
+#include <catboost/private/libs/options/enums.h>
+#include <library/threading/local_executor/local_executor.h>
+#include "util.h"
```
Comment (ek-ak) [inline catboost/libs/fstr/partial_dependence.h:7]: line should be in separate block above

File: catboost/libs/fstr/ya.make [hunk 14]
```diff
@@ -5,6 +5,7 @@ LIBRARY()
 SRCS(
     calc_fstr.cpp
     compare_documents.cpp
+    partial_dependence.cpp
```
Comment (ek-ak) [inline catboost/libs/fstr/ya.make:8]: lexicographical order

File: catboost/python-package/catboost/core.py [hunk 15]
```diff
@@ -2743,6 +2743,115 @@ def get_layout(go, feature, xaxis):
...
+    def plot_partial_dependence(self, data, features, plot=True, plot_file=None, thread_count=-1):
+        """
+        To use this function, you should install plotly.
+        data: numpy.ndarray or pandas.DataFrame or catboost.Pool
+        features: int, str, list<int>, tuple<int>, list<string>, tuple<string>
+            Float features to calculate partial dependence for. Number of features should be 1 or 2.
+        plot: bool
+            Plot predictions.
+        plot_file: str
+            Output file for plot predictions.
```
Comment (ek-ak) [inline catboost/python-package/catboost/core.py:2755]: thread_count count arg is missed

File: catboost/python-package/ut/medium/test.py [hunk 16]
```diff
@@ -8026,6 +8026,28 @@ def test_penalties_coefficient_work():
     assert any(model_without_feature_penalties.predict_proba(pool)[0] != model_with_feature_penalties.predict_proba(pool)[0])
 
 
+def test_partial_dependence_2d():
```
Comment (ek-ak) [inline catboost/python-package/ut/medium/test.py:8029]: try to use @pytest.mark.parametrize and merge this 2 tests

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by arcadia-devtools** [comment]
@ek-ak, internal review request created: [1258148](https://a.yandex-team.ru/review/1258148)

**Comment by arcadia-devtools** [comment]
Check [1258148](https://a.yandex-team.ru/review/1258148) failed: 23 build(s) broken, 14 test(s) broken

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@ek-ak, internal review request updated: [1258148](https://a.yandex-team.ru/review/1258148)

**Comment by arcadia-devtools** [comment]
Check [1258148](https://a.yandex-team.ru/review/1258148) failed: 23 build(s) broken, 4 test(s) broken

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@ek-ak, internal review request updated: [1258148](https://a.yandex-team.ru/review/1258148)

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@ek-ak, internal review request updated: [1258148](https://a.yandex-team.ru/review/1258148)

**Comment by arcadia-devtools** [comment]
Check [1258148](https://a.yandex-team.ru/review/1258148) passed

**Pull request merged**